orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java

```diff
@@ -0,0 +1,313 @@
+        .equals("`map`.`_key`", Type.LONG, 3L)
+        .equals("`listOfStruct`.`_elem`.`long`", Type.LONG, 4L)
+        .end()
+        .build();
+
+    SearchArgument actual = ExpressionToSearchArgument.visit(boundFilter, ORCSchemaUtil.convert(schema));
+    Assert.assertEquals(expected.toString(), actual.toString());
+  }
+
+  @Test
+  public void testSpecialCharacters() {
+    Schema schema = new Schema(
+        required(1, "col.with.dots", Types.StructType.of(
+            required(2, "inner.col.with.dots", Types.LongType.get())
+        )),
+        required(3, "colW!th$peci@lCh@rs", Types.LongType.get()),
+        required(4, "colWith`Quotes`", Types.LongType.get())
+    );
+
+    Expression expr = and(
+        equal("col.with.dots.inner.col.with.dots", 1),
+        and(equal("colW!th$peci@lCh@rs", 2), equal("colWith`Quotes`", 3))
+    );
+    Expression boundFilter = Binder.bind(schema.asStruct(), expr, true);
+    SearchArgument expected = SearchArgumentFactory.newBuilder()
+        .startAnd()
+        .equals("`col.with.dots`.`inner.col.with.dots`", Type.LONG, 1L)
+        .equals("`colW!th$peci@lCh@rs`", Type.LONG, 2L)
+        .equals("`colWith``Quotes```", Type.LONG, 3L)
+        .end()
+        .build();
+
+    SearchArgument actual = ExpressionToSearchArgument.visit(boundFilter, ORCSchemaUtil.convert(schema));
+    Assert.assertEquals(expected.toString(), actual.toString());
+  }
+
+  @Test
+  public void testEvolvedSchema() {
+    Schema fileSchema = new Schema(
+        required(1, "int", Types.IntegerType.get()),
+        optional(2, "long_to_be_dropped", Types.LongType.get())
+    );
+
+    Schema evolvedSchema = new Schema(
+        required(1, "int_renamed", Types.IntegerType.get()),
+        optional(3, "float_added", Types.FloatType.get())
+    );
+
+    TypeDescription readSchema = ORCSchemaUtil.buildOrcProjection(evolvedSchema, ORCSchemaUtil.convert(fileSchema));
+
+    Expression expr = equal("int_renamed", 1);
+    Expression boundFilter = Binder.bind(evolvedSchema.asStruct(), expr, true);
+    SearchArgument expected = SearchArgumentFactory.newBuilder()
+        .equals("`int`", Type.LONG, 1L)
+        .build();
+
+    SearchArgument actual = ExpressionToSearchArgument.visit(boundFilter, readSchema);
+    Assert.assertEquals(expected.toString(), actual.toString());
+
+    // for columns not in the file, buildOrcProjection will append field names with _r<ID>
+    // this will be passed down to ORC, but ORC will handle such cases and return a TruthValue during evaluation
+    expr = equal("float_added", 1);
+    boundFilter = Binder.bind(evolvedSchema.asStruct(), expr, true);
+    expected = SearchArgumentFactory.newBuilder()
+        .equals("`float_added_r3`", Type.FLOAT, 1.0)
+        .build();
+
+    actual = ExpressionToSearchArgument.visit(boundFilter, readSchema);
+    Assert.assertEquals(expected.toString(), actual.toString());
+  }
+
+  @Test
+  public void testIdToQuotedColumnName() {
```

Comment:
Can we move this to its own test suite? I don't think it should be mixed with the expression conversion tests.